File: Algorithm_77_Combinations.py
class Solution:
    def comb(self, select, k, cur, res):
        if cur==len(select):
            if k!=0: return
            thisRes=[]
            for i in range(len(select)):
                if select[i]:
                    thisRes.append(i+1)
            res.append(thisRes[:])
            return
        select[cur]=True
        Solution.comb(self,select,k-1,cur+1,res)
        select[cur]=False
        Solution.comb(self,select,k,cur+1,res)
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        select=[False for i in range(n)]
        res=[]
        Solution.comb(self,select,k,0,res)
        return res

sol=Solution()
res=sol.combine(4,2)
print (res)

File: Algorithm_77_Combinations.java
import java.util.List;
import java.util.ArrayList;

// 思路：递归回溯。用k记录剩余需要加入的数字的个数。当k == 0时，不需要再加入新的数字，则把curList加入res中
// cur表示上一层递归中被加入curList的数字，为防止重复，当前递归从cur + 1开始循环
// 循环变量i的最大值是n - k + 1，这是为了保证剩余的数字个数 >= k个

class Solution {

    private List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {
        helper(0, n, k, new ArrayList<Integer>());
        return res;
    }

    private void helper(int cur, int n, int k, List<Integer> curList) {
        if (k == 0) {
            res.add(new ArrayList<>(curList));
            return;
        }
        for (int i = cur + 1; i <= n - k + 1; i++) {
            curList.add(i);
            helper(i, n, k - 1, curList);
            curList.remove(curList.size() - 1);
        }
    }

}